# Cheat sheet

::: callout-warning
## Work in progress

This manual is a work in progress. Until it is complete, please continue using LINKDATAGEN from <https://bioinf.wehi.edu.au/software/linkdatagen/>.
:::

## Mandatory

Options are grouped. At least one option is required in each group.

-   Annotation (choose exactly one)
    -   `-annotFile <filename>`: File containing SNP annotation data.
    -   `-chip {1, 2, 3, 4, 5, 6, 7, 8}`: Chip (or HapMap) number \[note: must also use `--annotDir`\]

-   Call (choose exactly one)
    -    `-callDir <path to directory>`: Directory containing genotype data file(s).
    -    `-callFile <file path>`: File containing genotype data. 

-   `-data { a , i , m }`: Affymetrix (a) or Illumina SNP chip (i), or MPS/VCF data (m).
-   `-pedfile <filename>`: File containing pedigree information (pedfile).
-   Output (choose at least one and may specify multiple)
    - `-freq` Output allele frequencies of founders. 
    - `-prog { all | me | al | mo | pl | pr | cp | be | fe | re | fp }`: Format to output data in.
    - `-popHetTest { summary , verbose , perChr , perChrVerbose }`: Goodness-of-fit test of specified allele frequencies.
    - `-bestPopTest`: Identifies best population using allele frequencies.
-   Which Samples
    - `-whichSamplesFile <file path>`: File linking pedfile to genotype data file.
    - `-whichSamplesList <file path>`: File linking pedfile to genotype data file(s).

## Optional

```{=html}
<!---
The -actg line was removed from the following table as it is not functioning. 
This is a reminder to put it back if it works in the future.
| \-actg           | Not functioning.                                                     | Output data in ACTG format. Not functioning.       |
-->
```
| Option            | Values                                                               | Summary                                             |
|-------------------|-------------------------------|-----------------------|
| `-annotDir`       | \< path to directory \> \[ note: mandatory if using --chip \]        | Directory containing SNP annotation data file.      |
| `-binsize`        | { real number \>= 0.0 } \[ default = 0.3 \]                          | Non-overlapping bin size (in cM) to divide SNPs.    |
| `-crlmm`          | NO VALUE                                                             | Declare that data is in CRLMM format.               |
| `-fileKeepSNPs`   | \< filename \>                                                       | File containing SNPs to keep for selection.         |
| `-fileRemoveSNPs` | \< filename \>                                                       | File containing SNPs to exclude from selection.     |
| `-help`           | NO VALUE                                                             | Print help to screen.                               |
| `-minDist`        | { real number \>= 0.0 } \[ default = min { 0.2 , 0.5 \* binsize } \] | Minimum distance (cM) between selected SNPs.        |
| `-noX`            | NO VALUE                                                             | Declare that chr X data to be excluded from output. |
| `-outputDir`      | \< prefix \> \[ note: will be created or overwritten \]              | Prefix to output directory created.                 |
| `-pop`            | { ASW , CEU , CHB , CHD , GIH , JPT , LWK , MEX , MKK , TSI , YRI }  | Specify population allele frequencies (AF).         |
| `-popCol`         | { integer \>= 1 }                                                    | Specify column number of AF in annotation file.     |
| `-randomSNP`      | NO VALUE                                                             | Declare selection of a random SNP from each bin.    |
| `-regions`        | { #,chr#,#:######-######,chr#:######-######}                         | Specify regions of genome to analyse.               |
| `-regionsFile`    | \< filename \>                                                       | File containing regions of genome to analyse.       |
| `-removeAIS`      | NO VALUE                                                             |    Remove SNP markers that are absent from any Illumina input file. (use with caution) |
| `-removeWFHBS`    | { i , u }                                                            | Declare removal of SNPs with within-family HBS.     |
| `-keepME`         | NO VALUE                                                             | Declare keeping of SNPs with Mendelian errors.      |
| `-seed`           | { integer } \[ default = 12345 \]                                    | Seed to change "random" SNP selection.              |
