# Cheat sheet

::: callout-warning
## Work in progress

This manual is a work in progress. Until it is complete, please continue using LINKDATAGEN from <https://bioinf.wehi.edu.au/software/linkdatagen/>.
:::

## Mandatory

Options are grouped. At least one option is required in each group, and when a tick appears in the Mixed column multiple options from that group may be specified.

| Option              | Group   | Mixed | Values                                                               | Summary                                               |
|---------------|---------------|---------------|---------------|---------------|
| `-annotFile`        | Annot   | ❌    | \< filename \>                                                       | File containing SNP annotation data.                  |
| `-chip`             | Annot   | ❌    | { 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8} \[note: must also use \--annotDir\] | Chip (or HapMap) number.                              |
| `-callDir`          | Call    | ❌    | \< path to directory \>                                              | Directory containing genotype data file(s).           |
| `-callFile`         | Call    | ❌    | \< filename \>                                                       | File containing genotype data.                        |
| `-data`             | Data    | NA    | { a , i , m }                                                        | Affymetrix or Illumina SNP chip, or MPS data.         |
| `-pedfile`          | Pedfile | NA    | \< filename \>                                                       | File containing pedigree information (pedfile).       |
| `-freq`             | Output  | ✅    | NO VALUE \| Output allele                                            | Frequencies of founders.                              |
| `-prog`             | Output  | ✅    | { all \| me \| al \| mo \| pl \| pr \| cp \| be \| fe \| re \| fp }  | Format to output data in.                             |
| `-popHetTest`       | Output  | ✅    | { summary , verbose , perChr , perChrVerbose }                       | Goodness-of-fit test of specified allele frequencies. |
| `-bestPopTest`      | Output  | ✅    | NO VALUE \[note: mutually exclusive to the other 3 options\]         | Identifies best population using allele frequencies.  |
| `-whichSamplesFile` | WS      | ❌    | \< filename \>                                                       | File linking pedfile to genotype data file.           |
| `-whichSamplesList` | WS      | ❌    | \< filename \>                                                       | File linking pedfile to genotype data file(s).        |

## Others

```{=html}
<!---
The -actg line was removed from the following table as it is not functioning. 
This is a reminder to put it back if it works in the future.
| \-actg           | Not functioning.                                                     | Output data in ACTG format. Not functioning.       |
-->
```
| Option          | Values                                                               | Summary                                            |
|-------------------|------------------------------|-----------------------|
| -annotDir       | \< path to directory \> \[ note: mandatory if using --chip \]        | Directory containing SNP annotation data file.     |
| -binsize        | { real number \>= 0.0 } \[ default = 0.3 \]                          | Non-overlapping bin size (in cM) to divide SNPs.   |
| -crlmm          | NO VALUE                                                             | Declare that data is in CRLMM format.              |
| -fileKeepSNPs   | \< filename \>                                                       | File containing SNPs to keep for selection.        |
| -fileRemoveSNPs | \< filename \>                                                       | File containing SNPs to exclude from selection.    |
| -help           | NO VALUE                                                             | Print help to screen.                              |
| -minDist        | { real number \>= 0.0 } \[ default = min { 0.2 , 0.5 \* binsize } \] | Minimum distance (cM) between selected SNPs.       |
| -noX            | NO VALUE                                                             | Declare that chr X data to be exclude from output. |
| -outputDir      | \< prefix \> \[ note: will be created or overwritten \]              | Prefix to output directory created.                |
| -pop ,          | { ASW , CEU , CHB , CHD , GIH , JPT , LWK , MEX , MKK , TSI , YRI }  | Specify population allele frequencies (AF).        |
| -popCol         | { integer \>= 1 }                                                    | Specify column number of AF in annotation file.    |
| -randomSNP      | NO VALUE                                                             | Declare selection of a random SNP from each bin.   |
| -regions        | { #,chr#,#:######-######,chr#:######-######,... }                    | Specify regions of genome to analyse.              |
| -regionsFile    | \< filename \>                                                       | File containing regions of genome to analyse.      |
| -removeWFHBS    | { i , u }                                                            | Declare removal of SNPs with within-family HBS.    |
| -keepME         | NO VALUE                                                             | Declare keeping of SNPs with Mendelian errors.     |
| -seed           | { integer } \[ default = 12345 \]                                    | Seed to change "random" SNP selection.             |
