# Download and annotations {#sec-download}

::: callout-warning
## Work in progress

This manual is a work in progress. Until it is complete, please continue using LINKDATAGEN from <https://bioinf.wehi.edu.au/software/linkdatagen/>.
:::

To download LINKDATAGEN, go to the latest release page <https://github.com/bahlolab/linkdatagen/releases/latest> and download the `linkdatagen_<version>.zip file`.

It is assumed that [PERL](https://www.perl.org/) (5.8+) is installed. 
This is usually the case with Linux and macOS. 

## Annotations

In order to use LINKDATAGEN, you also need to download relevant annotations. The correct file to use will depend on the the platform.

### MPS and PLINK file format data

Download the [HapMap Phase II annotation files](https://bioinf-bucket.s3.ap-southeast-2.amazonaws.com/linkdatagen/annotation/mps/annotHapMap2U.txt.gz) (genome build b37). Last updated 3rd February 2016. Annotation for up to 4,031,388 SNPs for the four HapMap Phase II populations (CHB, CEU, JPT, YRI).\
Download the [HapMap Phase III annotation files](https://bioinf-bucket.s3.ap-southeast-2.amazonaws.com/linkdatagen/annotation/mps/annotHapMap3U.txt.gz) (genome build b37). Last updated 3rd February 2016. Annotation for up to 1,582,941 SNPs for the eleven HapMap Phase III populations.\
\
If you wish to perform genotype calling with [GATK's UnifiedGenotyper](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_genotyper_UnifiedGenotyper.php), then download the appropriate VCF file for [HapMap Phase II](https://bioinf-bucket.s3.ap-southeast-2.amazonaws.com/linkdatagen/annotation/mps/dbsnp_137.hg19_HapMap2L.vcf.gz) or [HapMap Phase III](https://bioinf-bucket.s3.ap-southeast-2.amazonaws.com/linkdatagen/annotation/mps/dbsnp_137.hg19_HapMap3L.vcf.gz) for the respective annotation file you plan to use.

#### Test data for MPS

We provide test data that was presented in the Smith KR et al (2011) paper. Last updated 3rd February 2016

Included in the test data: VCF files containing SNP genotypes at the location of HapMap Phase II SNPs: (i) Family A: Single affected individual - A-7.HapMapII.SNPs.vcf, recessive family, homozygosity mapping (ii) Family T: Single affected individual - T-1.HapMapII.SNPs.vcf, recessive family, homozygosity mapping (iii) Family M: Two affected siblings - M-3.HapMapII.SNPs.vcf and M-4.HapMapII.SNPs.vcf, dominant family.

SNP genotypes from Illumina genotyping arrays for the same individuals: (i) Family A: A-7.FinalReport.txt (ii) Family T: T-1.FinalReport.txt (iii) Family M: M-3.FinalReport.txt and M-4.FinalReport.txt

VCF files containing SNP genotypes at the location of genotyping array SNPs (for concordance checks): (i) Family A: A-7.IL610Q.SNPs.vcf (ii) Family T: T-1.IL610Q.SNPs.vcf (iii) Family M: M-3.ILOE.SNPs.vcf and M-4.ILOE.SNPs.vcf

### Affymetrix SNP chip data

Download the annotation files. These files are required by LINKDATAGEN and are generated by us from the (very large) Affymetrix annotation files and [HapMap Phase I, II and III data (unavailable).](http://www.hapmap.org/)\
For human genome build GRCh36/hg18 download the [Affymetrix build 36 annotation files](https://bioinf-bucket.s3.ap-southeast-2.amazonaws.com/linkdatagen/annotation/affymetrix/mappingfiles_affy_build36.tgz). Last updated 29th April 2013.\
For human genome build GRCh37/hg19 download the download the [Affymetrix build 37 annotation files](https://bioinf-bucket.s3.ap-southeast-2.amazonaws.com/linkdatagen/annotation/affymetrix/mappingfiles_affy_build37.tgz). Last updated 15th November 2012.\
\
For an example download the test data set [affymetrix_testdata.tar.gz](https://bioinf-bucket.s3.ap-southeast-2.amazonaws.com/linkdatagen/affymetrix_testdata.tar.gz). Last updated 15th November 2012.

### Illumina SNP chip data

The annotation files are required by LINKDATAGEN and are generated by us from the Illumina annotation files and [HapMap Phase I, II and III data (unavailable).](http://www.hapmap.org/) We support a range of Illumina chips including the 370Duo, 610Quad, Cyto12, OmniExpress, and 1M chips. These annotations are preferred when your SNP chip data is on these chips. For several other chips the best annotation file is suggested in the linkdatagen documentation.\
For human genome build GRCh37/hg19 download the [Illumina annotation files](https://bioinf-bucket.s3.ap-southeast-2.amazonaws.com/linkdatagen/annotation/annotIllumina.tgz). Last updated 3rd February 2016.

TODO: The MPS annotation files have been made to also work on Illumina SNP chip data. Investigate if these are usable (it's not well documented). It's possible specific SNP chip annotations have no advantage over the universal annotation files.
