---
nocite: |
  @bahlo2009linkagemapping
  @smith2011reducing
---
# Preface {.unnumbered}

::: callout-warning
## Work in progress

This manual is a work in progress. Until it is complete, please continue using LINKDATAGEN from <https://bioinf.wehi.edu.au/software/linkdatagen/>.
:::

::: {.content-visible when-format="html"}
::: callout-note 
## This manual is also available as a PDF or Word document. 
[linkdatagen-manual.pdf](linkdatagen-manual.pdf)

[linkdatagen-manual.docx](linkdatagen-manual.docx)
:::
:::

LINKDATAGEN is a PERL (5.8+) script that generates datasets for linkage analysis, relatedness checking, IBD and HBD inference and association analyses using genotypes generated from any of Affymetrix SNP chips, Illumina SNP chips or massively parallel sequencing (MPS; otherwise known as next generation sequencing or NGS) data.

LINKDATAGEN selects markers for linkage and association analyses and performs nuclear family Mendelian error detection. It also performs sex checks using both X and Y chromosome markers (if available). Linkdatagen supports all eleven HapMap populations (ASW, CEU, CHB, CHD, GIH, JPT, LWK, MEX, MKK, TSI and YRI) for the following platforms, unless otherwise specified:

-   Affymetrix: 50K Xba, 50K Hind, 250K Sty, 250K Nsp, 5.0, 6.0 and 500K Sty+Nsp
-   Illumina: 610Quad, Cyto12, Omni Express, 1M, plus others due to high overlap between Illumina platforms
-   MPS data: Hapmap2 (for CEU, CHB, JPT and YRI populations) and HapMap3
-   PLINK data (TODO: should this be included here?)

However, many other chips can be catered for (see -chip section). (TODO: insert cross-reference)

LINKDATAGEN creates output files for the linkage mapping software: ALLEGRO, MERLIN, MORGAN and PLINK, as well as for BEAGLE, FESTIM, PREST, fastPHASE and RELATE. Many of these programs are available through the Rockefeller website (<http://linkage.rockefeller.edu/soft>).

If you use LINKDATAGEN, please acknowledge by citing:

> Bahlo M, Bromhead CJ. Generating linkage mapping files from Affymetrix SNP chip data. *Bioinformatics* 2009;25(15):1961-2.


If you use the LINKDATAGEN -data m option along with vcf2linkdatagen.pl then please also cite:

> Smith KR, Bromhead CJ, Hildebrand MS, Shearer AE, Lockhart PJ, Najmabadi H, Leventer RJ, McGillivray G, Amor DJ, Smith RJ, Bahlo M. Reducing the exome search space for Mendelian diseases using genetic linkage analysis of exome genotypes. *Genome Biology* 2011;12:R85.

## History

-   The first incarnation of linkdatagen.pl was only able to process genotypes from Affymetrix SNP chips.
-   Subsequently, linkdatagen.pl was renamed linkdatagen_affy.pl. Separate scripts were developed for Illumina SNP chip genotypes (linkdatagen_illumina.pl) and MPS genotypes (linkdatagen_mps.pl and companion script vcf2linkdatagen.pl).
-   As of the 15th May 2012, the three linkdatagen scripts have been combined into a single script named linkdatagen.pl. The type of genotypes being processed is indicated by the -data option ('a' for Affymetrix SNP chip data, 'i' for Illumina SNP chip data or 'm' for SNP genotypes from MPS data). vcf2linkdatagen.pl remains a separate script that must be run before using linkdatagen.pl with the -data m option.
-   On 3rd February 2016, a new annotation format was released for MPS and Illumina SNP chip data that can be used across both formats. Where annotation files for new Illumina SNP chips have not been released the new annotation file may be of benefit. MPS genotypes can now be called with GATK Unified Genotyper.
-   On XX YYYYYY 2022 the LINKDATAGEN manual was written as a website and PDF.
-   On XX YYYYYY 2022 the following updates were made.
    -   GATK HaplotypeCaller support was added.
    -   Multi-sample VCFs were supported.
    -   PLINK format files could be used after conversion to VCF.
    -   hg38 annotations were released.
